File: android-native/app/src/main/java/com/example/kpkn/screens/workout/WorkoutFeedbackModels.kt
package com.example.kpkn.screens.workout

import com.example.kpkn.data.exercises.resolveCatalogExerciseInfoInIndex
import com.example.kpkn.data.models.CompletedSet
import com.example.kpkn.data.models.DropSetData
import com.example.kpkn.data.models.ExerciseMuscleInfo
import com.example.kpkn.data.models.MuscleRole
import com.example.kpkn.data.models.PlanDeviation
import com.example.kpkn.data.models.PlanDeviationType
import com.example.kpkn.data.models.PostExerciseFeedback
import com.example.kpkn.data.models.PostSessionFeedback
import com.example.kpkn.data.models.MuscleFeedbackEntry
import com.example.kpkn.data.models.RestPauseData
import com.example.kpkn.data.models.ExerciseSet
import com.example.kpkn.data.models.HistoryColorV2
import com.example.kpkn.data.models.LoadModeV2
import com.example.kpkn.data.models.IntensityMode
import com.example.kpkn.data.models.WorkoutLog
import com.example.kpkn.data.models.effectiveRepRange
import com.example.kpkn.domain.auge.AugeFatigueEngine
import com.example.kpkn.domain.auge.getAugeMusclePillarId
import com.example.kpkn.domain.training.VolumeCalculator
import java.time.LocalDate

data class SetAdvancedFeedback(
    val rir: Int? = null,
    val reachedFailure: Boolean = false,
    val isFailedSet: Boolean = false,
    val failureReason: String? = null,
    val executionError: Boolean = false,
    val isPartial: Boolean = false,
    val partialReps: Int? = null,
    val dropSets: List<DropSetData> = emptyList(),
    val restPauses: List<RestPauseData> = emptyList(),
    val skipped: Boolean = false,
    val superSetWithExerciseId: String? = null,
    val isWarmup: Boolean = false,
    val actualIntensityMode: IntensityMode? = null,
    val actualIntensityValue: Double? = null,
    /** Nullable live override: null follows the planned set, false disables planned AMRAP. */
    val amrapOverride: Boolean? = null,
    val amrapMinimumReps: Int? = null,
    val timerElapsedSeconds: Int? = null,
    val timerTargetSeconds: Int? = null,
    val rom: Int? = null,
    val assistedReps: Int? = null,
)

internal fun resolveAmrapActive(
    plannedSet: ExerciseSet?,
    requestedOverride: Boolean,
    explicitOverride: Boolean? = null,
): Boolean {
    val override = explicitOverride ?: requestedOverride.takeIf { it }
    return override ?: (plannedSet?.isAmrap == true || plannedSet?.intensityMode == IntensityMode.AMRAP)
}

sealed interface PostExerciseFeedbackTarget {
    data class Single(val exerciseId: String) : PostExerciseFeedbackTarget
    data class SupersetGroup(
        val groupId: String,
        val exerciseIds: List<String>,
    ) : PostExerciseFeedbackTarget
}

internal fun backfillCompletedSetIntensityFromPostExerciseFeedback(
    completedSets: Map<String, CompletedSet>,
    feedback: PostExerciseFeedback,
): Map<String, CompletedSet> {
    val perceived = feedback.perceivedIntensityRpe?.coerceIn(1.0, 10.0)
    if (perceived == null && !feedback.perceivedFailure) return completedSets
    val mode = if (feedback.perceivedFailure) IntensityMode.FAILURE else IntensityMode.RPE
    val value = if (feedback.perceivedFailure) 10.0 else perceived
    return completedSets.mapValues { (key, set) ->
        val belongsToExercise = key.startsWith("${feedback.exerciseId}_")
        val hasRecordedIntensity = set.actualIntensityValue != null ||
            set.actualIntensityMode != null ||
            set.rpe != null ||
            set.rir != null ||
            set.isFailure
        if (!belongsToExercise || hasRecordedIntensity) {
            set
        } else {
            set.copy(
                rpe = if (feedback.perceivedFailure) null else value,
                isFailure = feedback.perceivedFailure,
                actualIntensityMode = mode,
                actualIntensityValue = value,
            )
        }
    }
}

internal fun backfillCompletedSetIntensityFromPostExerciseFeedbacks(
    completedSets: Map<String, CompletedSet>,
    feedbacks: List<PostExerciseFeedback>,
): Map<String, CompletedSet> = feedbacks.fold(completedSets) { current, feedback ->
    backfillCompletedSetIntensityFromPostExerciseFeedback(current, feedback)
}

data class SessionClosingFeedback(
    val overallFatigue: Int,
    val systemAdjustment: Int,
    val muscularAdjustment: Int,
    val structureAdjustment: Int,
    val discomforts: List<String>,
    val clarityRating: Int = 5,           // 1–10: mental clarity / freshness at end
    val environmentTags: List<String> = emptyList(), // e.g. "buen sueño", "estresado"
    val finalNeuralBattery: Int? = null,
    val finalSpinalBattery: Int? = null,
    val finalMuscularBattery: Int? = null,
    val finalMuscleBatteries: Map<String, Int> = emptyMap(),
    /** True when the user dragged/edited the neural ring vs engine seed. */
    val neuralEdited: Boolean = false,
    /** True when the user dragged/edited the spinal ring vs engine seed. */
    val spinalEdited: Boolean = false,
    /** True when the user edited any per-muscle slider vs engine seed. */
    val musclesEdited: Boolean = false,
    val additionalDiscomfortNote: String? = null,
    val stillPresentDiscomfortIds: List<String> = emptyList(),
)

data class ExerciseHistoryEntry(
    val date: String,
    val sets: List<CompletedSet>,
    val e1rm: Double?,
    val tag: String? = null,  // tag usado en esa sesión para ese ejercicio
    val notes: String? = null,
    val latestHistoryColor: HistoryColorV2? = null,
    val latestMetricType: String? = null,
    val latestMetricValue: Double? = null,
)

data class WeightSuggestion(
    val suggestedWeight: Double,
    val reason: String,
    val suggestedLoadMode: LoadModeV2? = null,
)

data class WorkoutLoadSuggestionUi(
    val suggestedWeight: Double,
    val originalWeight: Double,
    val isRecalculated: Boolean = false,
    val reason: String,
    val source: WorkoutLoadSuggestionSource = WorkoutLoadSuggestionSource.PROGRAM,
    val suggestedLoadMode: LoadModeV2? = null,
)

enum class WorkoutLoadSuggestionSource {
    PROGRAM,
    HISTORY,
    SESSION_ERM,
    MANUAL_BASE,
}

object WorkoutPlanDeviationSupport {
    fun detect(
        exerciseId: String,
        exerciseName: String,
        setIdx: Int,
        plannedSet: ExerciseSet,
        actualWeight: Double,
        actualReps: Int,
        advanced: SetAdvancedFeedback,
        suggestedWeight: Double?,
        amrapMinimumReps: Int? = advanced.amrapMinimumReps,
    ): List<PlanDeviation> {
        val deviations = mutableListOf<PlanDeviation>()

        val targetWeight = suggestedWeight ?: 0.0
        if (targetWeight > 0 && actualWeight > 0) {
            val ratio = actualWeight / targetWeight
            if (ratio > 1.15) deviations.add(PlanDeviation(exerciseId, exerciseName, setIdx, PlanDeviationType.WEIGHT_HIGH, "+${"%.0f".format((ratio - 1) * 100)}% del sugerido"))
            else if (ratio < 0.85) deviations.add(PlanDeviation(exerciseId, exerciseName, setIdx, PlanDeviationType.WEIGHT_LOW, "-${"%.0f".format((1 - ratio) * 100)}% del sugerido"))
        }

        val targetRange = plannedSet.effectiveRepRange()
        val amrapActive = resolveAmrapActive(
            plannedSet = plannedSet,
            requestedOverride = false,
            explicitOverride = advanced.amrapOverride,
        )
        if (targetRange != null && actualReps > 0) {
            if (amrapActive) {
                val minimum = amrapMinimumReps ?: targetRange.min
                if (actualReps < minimum) {
                    deviations.add(
                        PlanDeviation(
                            exerciseId,
                            exerciseName,
                            setIdx,
                            PlanDeviationType.AMRAP_BELOW_MINIMUM,
                            "$actualReps vs mínimo AMRAP $minimum",
                        ),
                    )
                }
            } else if (actualReps > targetRange.max + 3) {
                deviations.add(PlanDeviation(exerciseId, exerciseName, setIdx, PlanDeviationType.REPS_HIGH, "$actualReps vs ${targetRange.format()} objetivo"))
            } else if (actualReps < targetRange.min - 3) {
                deviations.add(PlanDeviation(exerciseId, exerciseName, setIdx, PlanDeviationType.REPS_LOW, "$actualReps vs ${targetRange.format()} objetivo"))
            }
        }

        if (advanced.reachedFailure && !plannedSet.isFailure) deviations.add(PlanDeviation(exerciseId, exerciseName, setIdx, PlanDeviationType.UNPLANNED_FAILURE, "Fallo no programado"))
        if (advanced.dropSets.isNotEmpty() && plannedSet.dropSets.isEmpty()) deviations.add(PlanDeviation(exerciseId, exerciseName, setIdx, PlanDeviationType.UNPLANNED_DROPSET, "Dropset no programado"))
        if (advanced.restPauses.isNotEmpty() && plannedSet.restPauses.isEmpty()) deviations.add(PlanDeviation(exerciseId, exerciseName, setIdx, PlanDeviationType.UNPLANNED_REST_PAUSE, "Rest-pause no programado"))

        return deviations
    }
}

fun applyAdvancedFeedback(
    base: CompletedSet,
    advanced: SetAdvancedFeedback,
): CompletedSet {
    return base.copy(
        rir = advanced.rir,
        isFailure = advanced.reachedFailure,
        isFailedSet = advanced.isFailedSet || advanced.executionError,
        failureReason = advanced.failureReason ?: if (advanced.executionError) "execution_error" else null,
        isPartial = advanced.isPartial,
        partialReps = advanced.partialReps,
        dropSets = advanced.dropSets,
        restPauses = advanced.restPauses,
        skipped = advanced.skipped,
        superSetWithExerciseId = advanced.superSetWithExerciseId,
        isWarmup = advanced.isWarmup,
        // A set without planned intensity must remain unlabelled in the live
        // record. Only explicit feedback may add an actual mode/value later.
        actualIntensityMode = advanced.actualIntensityMode ?: base.actualIntensityMode,
        actualIntensityValue = advanced.actualIntensityValue ?: base.actualIntensityValue,
        rom = advanced.rom,
        assistedReps = advanced.assistedReps,
    )
}

fun calculateUnifiedSessionEffortSignal(
    sets: List<CompletedSet>,
): Double {
    val effectiveSets = sets.filter { set ->
        !set.isWarmup && AugeFatigueEngine.isSetEffective(set)
    }
    if (effectiveSets.isEmpty()) return 7.0

    // getEffectiveRPE already includes technique bonuses — do not double-count
    return effectiveSets
        .map { set -> AugeFatigueEngine.getEffectiveRPE(set).coerceIn(1.0, 12.0) }
        .average()
        .coerceIn(1.0, 12.0)
}

/** Maps a 1–10 technical quality score to the 1–5 scale expected by calculateTechniquePenalty. */
fun technicalQuality10ToPenaltyScale(technicalQuality: Int): Int {
    val q = technicalQuality.coerceIn(1, 10)
    return (1 + ((q - 1) * 4) / 9).coerceIn(1, 5)
}

@Suppress("unused")
fun mapWorkoutToPostSessionFeedback(
    log: WorkoutLog,
    postExerciseFeedback: List<PostExerciseFeedback>,
    exerciseDbById: Map<String, ExerciseMuscleInfo>,
): PostSessionFeedback {
    val grouped = postExerciseFeedback.groupBy { feedback ->
        val exercise = log.completedExercises.find { it.exerciseId == feedback.exerciseId }
        val dbInfo = exercise?.let {
            resolveCatalogExerciseInfoInIndex(
                index = exerciseDbById,
                catalogConfigurationId = it.catalogConfigurationId,
                exerciseDbId = it.exerciseDbId,
                exerciseId = it.exerciseId,
                exerciseName = it.exerciseName,
            )
        }
        val primary = (exercise?.effectiveMuscles?.takeIf { it.isNotEmpty() } ?: dbInfo?.involvedMuscles.orEmpty())
            .firstOrNull { it.role == MuscleRole.PRIMARY }
        if (primary != null) {
            val canonical = VolumeCalculator.normalizeCanonicalMuscleGroup(primary.muscle, primary.emphasis)
            getAugeMusclePillarId(canonical, primary.emphasis)
        } else feedback.exerciseName
    }

    val feedbackByMuscle = grouped.mapValues { (_, list) ->
        val avgDemand = 7
        val avgTech = if (list.isEmpty()) 8 else list.map { it.technicalQuality }.average().toInt().coerceIn(1, 10)
        val hasJointPain = list.any { it.discomfortIds.any { d -> d != "none" } }
        MuscleFeedbackEntry(
            doms = (11 - avgTech).coerceIn(1, 5),
            jointPain = hasJointPain,
            strengthCapacity = avgDemand,
            notes = list.flatMap { it.discomfortIds }.filter { it != "none" }.distinct().joinToString().takeIf { it.isNotBlank() } ?: "",
        )
    }

    val cnsRecovery = 7

    return PostSessionFeedback(
        logId = log.id,
        date = LocalDate.now().toString(),
        cnsRecovery = cnsRecovery,
        muscleFeedback = feedbackByMuscle,
    )
}

data class SetAutoRegulation(
    val exerciseId: String,
    val nextSetIdx: Int,
    val adjustmentFactor: Double,
    val adjustedWeight: Double,
    val reason: String,
)

object WorkoutAutoRegulation {
    private const val MIN_FACTOR = 0.60
    private const val MAX_FACTOR = 1.10

    fun computeAdjustmentFactor(
        weightedDrainPct: Double,
        effectiveRpe: Double,
        reachedFailure: Boolean,
        isFailedSet: Boolean,
        isPartial: Boolean,
        sessionProgress: Double,
    ): Double {
        var factor = 1.0

        val drainFactor = when {
            weightedDrainPct >= 10.0 -> -0.10
            weightedDrainPct >= 7.0 -> -0.07
            weightedDrainPct >= 4.5 -> -0.04
            weightedDrainPct >= 2.5 -> -0.02
            weightedDrainPct >= 1.0 -> -0.01
            else -> 0.0
        }
        factor += drainFactor

        val rpeFactor = when {
            effectiveRpe >= 11.0 -> -0.08
            effectiveRpe >= 10.5 -> -0.06
            effectiveRpe >= 10.0 -> -0.04
            effectiveRpe >= 9.5 -> -0.02
            effectiveRpe >= 8.5 -> -0.01
            effectiveRpe <= 7.0 -> +0.02
            else -> 0.0
        }
        factor += rpeFactor

        if (reachedFailure) factor -= 0.05
        if (isFailedSet) factor -= 0.03
        if (isPartial) factor -= 0.02

        val sessionFactor = when {
            sessionProgress >= 0.80 -> -0.03
            sessionProgress >= 0.60 -> -0.02
            sessionProgress >= 0.40 -> -0.01
            else -> 0.0
        }
        factor += sessionFactor

        return factor.coerceIn(MIN_FACTOR, MAX_FACTOR)
    }

    @Suppress("UNUSED_PARAMETER")
    fun buildReason(
        factor: Double,
        weightedDrainPct: Double,
        effectiveRpe: Double,
        reachedFailure: Boolean,
    ): String {
        val parts = mutableListOf<String>()
        if (reachedFailure) parts.add("Fallo")
        if (weightedDrainPct >= 5.0) parts.add("Fatiga acumulada")
        val detail = parts.joinToString(" · ")
        return when {
            factor < 0.95 -> buildString {
                append("Ajuste de carga")
                if (detail.isNotBlank()) append(" · $detail")
                append(" · −${((1 - factor) * 100).toInt()}%")
            }
            factor > 1.02 -> "Ajuste de carga · Recuperación buena · +${((factor - 1) * 100).toInt()}%"
            else -> "Sin ajuste de carga"
        }
    }

    private const val READINESS_REDUCTION_THRESHOLD = 70
    private const val READINESS_SEVERE_THRESHOLD = 50
    private const val READINESS_MIN_FACTOR = 0.70
    private const val READINESS_SEVERE_FACTOR = 0.85
    private const val READINESS_RECOVERY_FACTOR = 1.05

    fun computeReadinessAdjustmentFactor(
        readinessNeural: Int?,
        readinessSpinal: Int?,
        readinessMuscular: Int?,
        readinessPerMuscle: Map<String, Int>?,
        involvedMuscleIds: List<String>,
    ): Double {
        val perMuscleValues = readinessPerMuscle?.let { overrides ->
            val relevant = involvedMuscleIds.mapNotNull { overrides[it] }
            if (relevant.isNotEmpty()) relevant.min() else null
        }
        val relevantReadiness = listOfNotNull(
            readinessNeural?.toDouble(),
            readinessSpinal?.toDouble(),
            readinessMuscular?.toDouble(),
            perMuscleValues?.toDouble(),
        ).minOrNull()

        val readiness = (relevantReadiness ?: return 1.0).coerceIn(0.0, 100.0)

        return when {
            readiness < READINESS_SEVERE_THRESHOLD -> {
                val t = (readiness / READINESS_SEVERE_THRESHOLD).coerceIn(0.0, 1.0)
                (READINESS_MIN_FACTOR + (READINESS_SEVERE_FACTOR - READINESS_MIN_FACTOR) * t)
                    .coerceIn(READINESS_MIN_FACTOR, READINESS_SEVERE_FACTOR)
            }
            readiness < READINESS_REDUCTION_THRESHOLD -> {
                val t = ((readiness - READINESS_SEVERE_THRESHOLD) /
                    (READINESS_REDUCTION_THRESHOLD - READINESS_SEVERE_THRESHOLD)).coerceIn(0.0, 1.0)
                (READINESS_SEVERE_FACTOR + (1.0 - READINESS_SEVERE_FACTOR) * t)
                    .coerceIn(READINESS_SEVERE_FACTOR, 1.0)
            }
            readiness >= 90 -> READINESS_RECOVERY_FACTOR
            else -> 1.0
        }
    }

    fun buildReadinessReason(factor: Double, readinessValue: Int?): String {
        if (factor == 1.0) return ""
        return when {
            factor < 0.85 && readinessValue != null -> "Readiness ${readinessValue} · −${((1 - factor) * 100).toInt()}%"
            factor < 1.0 -> "Readiness ${readinessValue ?: "baja"}"
            factor > 1.0 -> "Readiness ${readinessValue ?: "alta"} · +${((factor - 1) * 100).toInt()}%"
            else -> ""
        }
    }
}

File: ios-native/KPKNFit/KPKNFit/Presentation/Screens/Workout/WorkoutFeedbackModels.swift
import Foundation

struct SetAdvancedFeedback {
    var rir: Int? = nil
    var reachedFailure: Bool = false
    var isFailedSet: Bool = false
    var failureReason: String? = nil
    var executionError: Bool = false
    var isPartial: Bool = false
    var partialReps: Int? = nil
    var dropSets: [DropSetData] = []
    var restPauses: [RestPauseData] = []
    var skipped: Bool = false
    var superSetWithExerciseId: String? = nil
    var isWarmup: Bool = false
    var actualIntensityMode: IntensityMode? = nil
    var actualIntensityValue: Double? = nil
    var timerElapsedSeconds: Int? = nil
    var timerTargetSeconds: Int? = nil
    var rom: Int? = nil
    var assistedReps: Int? = nil
}

struct PostExerciseFeedback {
    let exerciseId: String
    let exerciseDbId: String?
    let canonicalExerciseId: String?
    let exerciseName: String
    let technicalQuality: Int
    let discomfortIds: [String]
    let notes: String?
    let perceivedIntensityRpe: Double?
    let perceivedFailure: Bool

    init(
        exerciseId: String,
        exerciseDbId: String? = nil,
        canonicalExerciseId: String? = nil,
        exerciseName: String,
        technicalQuality: Int,
        discomfortIds: [String] = [],
        notes: String? = nil,
        perceivedIntensityRpe: Double? = nil,
        perceivedFailure: Bool = false
    ) {
        self.exerciseId = exerciseId
        self.exerciseDbId = exerciseDbId
        self.canonicalExerciseId = canonicalExerciseId
        self.exerciseName = exerciseName
        self.technicalQuality = technicalQuality
        self.discomfortIds = discomfortIds
        self.notes = notes
        self.perceivedIntensityRpe = perceivedIntensityRpe
        self.perceivedFailure = perceivedFailure
    }
}

enum PostExerciseFeedbackTarget {
    case single(exerciseId: String)
    case supersetGroup(groupId: String, exerciseIds: [String])
}

func backfillCompletedSetIntensityFromPostExerciseFeedback(
    completedSets: [String: CompletedSet],
    feedback: PostExerciseFeedback
) -> [String: CompletedSet] {
    let perceived = feedback.perceivedIntensityRpe.flatMap { $0.clamped(to: 1.0...10.0) }
    if perceived == nil && !feedback.perceivedFailure { return completedSets }
    let mode: IntensityMode = feedback.perceivedFailure ? .failure : .rpe
    let value: Double = feedback.perceivedFailure ? 10.0 : (perceived ?? 10.0)
    return completedSets.mapValues { set in
        let belongsToExercise = set.key.hasPrefix("\(feedback.exerciseId)_")
        let hasRecordedIntensity = set.actualIntensityValue != nil ||
            set.actualIntensityMode != nil ||
            set.rpe != nil ||
            set.rir != nil ||
            set.isFailure
        guard belongsToExercise, !hasRecordedIntensity else { return set }
        return set.copyWith(
            rpe: feedback.perceivedFailure ? nil : value,
            isFailure: feedback.perceivedFailure,
            actualIntensityMode: mode,
            actualIntensityValue: value
        )
    }
}

func backfillCompletedSetIntensityFromPostExerciseFeedbacks(
    completedSets: [String: CompletedSet],
    feedbacks: [PostExerciseFeedback]
) -> [String: CompletedSet] {
    return feedbacks.reduce(completedSets) { current, feedback in
        backfillCompletedSetIntensityFromPostExerciseFeedback(completedSets: current, feedback: feedback)
    }
}

struct SessionClosingFeedback {
    let overallFatigue: Int
    let systemAdjustment: Int
    let muscularAdjustment: Int
    let structureAdjustment: Int
    let discomforts: [String]
    var clarityRating: Int = 5
    var environmentTags: [String] = []
    var finalNeuralBattery: Int? = nil
    var finalSpinalBattery: Int? = nil
    var finalMuscularBattery: Int? = nil
    var finalMuscleBatteries: [String: Int] = [:]
    var additionalDiscomfortNote: String? = nil
    var stillPresentDiscomfortIds: [String] = []
}

struct ExerciseHistoryEntry {
    let date: String
    let sets: [CompletedSet]
    let e1rm: Double?
    let tag: String?
    let latestHistoryColor: HistoryColorV2?
    let latestMetricType: String?
    let latestMetricValue: Double?
}

struct WeightSuggestion {
    let suggestedWeight: Double
    let reason: String
    let suggestedLoadMode: LoadModeV2?
}

struct WorkoutLoadSuggestionUi {
    let suggestedWeight: Double
    let originalWeight: Double
    var isRecalculated: Bool = false
    let reason: String
    let source: WorkoutLoadSuggestionSource
    let suggestedLoadMode: LoadModeV2?
}

enum WorkoutLoadSuggestionSource {
    case program
    case history
    case sessionErm
    case manualBase
}

struct WorkoutPlanDeviationSupport {
    static func detect(
        exerciseId: String,
        exerciseName: String,
        setIdx: Int,
        plannedSet: ExerciseSet,
        actualWeight: Double,
        actualReps: Int,
        advanced: SetAdvancedFeedback,
        suggestedWeight: Double?
    ) -> [PlanDeviation] {
        var deviations: [PlanDeviation] = []
        let targetWeight = suggestedWeight ?? 0.0
        if targetWeight > 0 && actualWeight > 0 {
            let ratio = actualWeight / targetWeight
            if ratio > 1.15 {
                deviations.append(PlanDeviation(
                    exerciseId: exerciseId, exerciseName: exerciseName, setIdx: setIdx,
                    type: .weightHigh, detail: String(format: "+%.0f%% del sugerido", (ratio - 1) * 100)
                ))
            } else if ratio < 0.85 {
                deviations.append(PlanDeviation(
                    exerciseId: exerciseId, exerciseName: exerciseName, setIdx: setIdx,
                    type: .weightLow, detail: String(format: "-%.0f%% del sugerido", (1 - ratio) * 100)
                ))
            }
        }
        let targetReps = plannedSet.targetReps
        if let targetReps = targetReps, targetReps > 0, actualReps > 0 {
            if actualReps > targetReps + 3 {
                deviations.append(PlanDeviation(
                    exerciseId: exerciseId, exerciseName: exerciseName, setIdx: setIdx,
                    type: .repsHigh, detail: "\(actualReps) vs \(targetReps) objetivo"
                ))
            } else if actualReps < targetReps - 3 {
                deviations.append(PlanDeviation(
                    exerciseId: exerciseId, exerciseName: exerciseName, setIdx: setIdx,
                    type: .repsLow, detail: "\(actualReps) vs \(targetReps) objetivo"
                ))
            }
        }
        if advanced.reachedFailure && !plannedSet.isFailure {
            deviations.append(PlanDeviation(
                exerciseId: exerciseId, exerciseName: exerciseName, setIdx: setIdx,
                type: .unplannedFailure, detail: "Fallo no programado"
            ))
        }
        if !advanced.dropSets.isEmpty && plannedSet.dropSets.isEmpty {
            deviations.append(PlanDeviation(
                exerciseId: exerciseId, exerciseName: exerciseName, setIdx: setIdx,
                type: .unplannedDropset, detail: "Dropset no programado"
            ))
        }
        if !advanced.restPauses.isEmpty && plannedSet.restPauses.isEmpty {
            deviations.append(PlanDeviation(
                exerciseId: exerciseId, exerciseName: exerciseName, setIdx: setIdx,
                type: .unplannedRestPause, detail: "Rest-pause no programado"
            ))
        }
        return deviations
    }
}

func applyAdvancedFeedback(base: CompletedSet, advanced: SetAdvancedFeedback) -> CompletedSet {
    return base.copyWith(
        rir: advanced.rir,
        isFailure: advanced.reachedFailure,
        isFailedSet: advanced.isFailedSet || advanced.executionError,
        failureReason: advanced.failureReason ?? (advanced.executionError ? "execution_error" : nil),
        isPartial: advanced.isPartial,
        partialReps: advanced.partialReps,
        dropSets: advanced.dropSets,
        restPauses: advanced.restPauses,
        skipped: advanced.skipped,
        superSetWithExerciseId: advanced.superSetWithExerciseId,
        isWarmup: advanced.isWarmup,
        actualIntensityMode: advanced.actualIntensityMode,
        actualIntensityValue: advanced.actualIntensityValue,
        rom: advanced.rom,
        assistedReps: advanced.assistedReps
    )
}

func calculateUnifiedSessionEffortSignal(sets: [CompletedSet]) -> Double {
    let effectiveSets = sets.filter { set in
        !set.isWarmup && AugeFatigueEngine.isSetEffective(set)
    }
    guard !effectiveSets.isEmpty else { return 7.0 }
    let signals = effectiveSets.map { set -> Double in
        var signal = AugeFatigueEngine.getEffectiveRPE(set)
        if set.isFailure { signal += 0.6 }
        if !set.dropSets.isEmpty { signal += 0.4 }
        if !set.restPauses.isEmpty { signal += 0.5 }
        return signal.clamped(to: 1.0...12.0)
    }
    let avg = signals.reduce(0.0, +) / Double(signals.count)
    return avg.clamped(to: 1.0...12.0)
}

func mapWorkoutToPostSessionFeedback(
    log: WorkoutLog,
    postExerciseFeedback: [PostExerciseFeedback],
    exerciseDbById: [String: ExerciseMuscleInfo]
) -> PostSessionFeedback {
    let grouped = Dictionary(grouping: postExerciseFeedback) { feedback -> String in
        let exercise = log.completedExercises.first { $0.exerciseId == feedback.exerciseId }
        let dbInfo = exercise.flatMap { exerciseDbById[$0.exerciseDbId ?? $0.exerciseId] }
        if let primary = dbInfo?.involvedMuscles.first(where: { $0.role == .primary }) {
            let canonical = VolumeCalculator.normalizeCanonicalMuscleGroup(primary.muscle, primary.emphasis)
            return getAugeMuscleDisplayId(canonical: canonical, emphasis: primary.emphasis)
        }
        return feedback.exerciseName
    }
    let feedbackByMuscle = grouped.mapValues { list in
        let avgTech = list.isEmpty ? 8 : list.map { $0.technicalQuality }.reduce(0, +) / list.count
        let hasJointPain = list.contains { $0.discomfortIds.contains { $0 != "none" } }
        return MuscleFeedbackEntry(
            doms: (11 - avgTech).clamped(to: 1...5),
            jointPain: hasJointPain,
            strengthCapacity: 7,
            notes: list.flatMap { $0.discomfortIds }.filter { $0 != "none" }.joined(separator: ", ")
        )
    }
    return PostSessionFeedback(
        logId: log.id,
        date: Date().formatted(.iso8601),
        cnsRecovery: 7,
        muscleFeedback: feedbackByMuscle
    )
}

struct SetAutoRegulation {
    let exerciseId: String
    let nextSetIdx: Int
    let adjustmentFactor: Double
    let adjustedWeight: Double
    let reason: String
}

struct WorkoutAutoRegulation {
    private static let minFactor = 0.60
    private static let maxFactor = 1.10

    static func computeAdjustmentFactor(
        weightedDrainPct: Double,
        effectiveRpe: Double,
        reachedFailure: Bool,
        isFailedSet: Bool,
        isPartial: Bool,
        sessionProgress: Double
    ) -> Double {
        var factor = 1.0
        let drainFactor: Double = {
            switch weightedDrainPct {
            case 10.0...: return -0.10
            case 7.0..<10.0: return -0.07
            case 4.5..<7.0: return -0.04
            case 2.5..<4.5: return -0.02
            case 1.0..<2.5: return -0.01
            default: return 0.0
            }
        }()
        factor += drainFactor
        let rpeFactor: Double = {
            switch effectiveRpe {
            case 11.0...: return -0.08
            case 10.5..<11.0: return -0.06
            case 10.0..<10.5: return -0.04
            case 9.5..<10.0: return -0.02
            case 8.5..<9.5: return -0.01
            case ..<7.5: return +0.02
            default: return 0.0
            }
        }()
        factor += rpeFactor
        if reachedFailure { factor -= 0.05 }
        if isFailedSet { factor -= 0.03 }
        if isPartial { factor -= 0.02 }
        let sessionFactor: Double = {
            switch sessionProgress {
            case 0.80...: return -0.03
            case 0.60..<0.80: return -0.02
            case 0.40..<0.60: return -0.01
            default: return 0.0
            }
        }()
        factor += sessionFactor
        return factor.clamped(to: minFactor...maxFactor)
    }

    static func buildReason(factor: Double, weightedDrainPct: Double, effectiveRpe: Double, reachedFailure: Bool) -> String {
        var parts: [String] = []
        if reachedFailure { parts.append("Fallo") }
        if weightedDrainPct >= 5.0 { parts.append("Fatiga acumulada") }
        let detail = parts.joined(separator: " · ")
        if factor < 0.95 {
            var result = "Ajuste de carga"
            if !detail.isEmpty { result += " · \(detail)" }
            result += " · −\((Int((1 - factor) * 100)))%"
            return result
        } else if factor > 1.02 {
            return "Ajuste de carga · Recuperación buena · +\((Int((factor - 1) * 100)))%"
        }
        return "Sin ajuste de carga"
    }

    private static let readinessReductionThreshold = 70
    private static let readinessSevereThreshold = 50
    private static let readinessMinFactor = 0.70
    private static let readinessSevereFactor = 0.85
    private static let readinessRecoveryFactor = 1.05

    static func computeReadinessAdjustmentFactor(
        readinessNeural: Int?,
        readinessSpinal: Int?,
        readinessMuscular: Int?,
        readinessPerMuscle: [String: Int]?,
        involvedMuscleIds: [String]
    ) -> Double {
        let perMuscleValues = readinessPerMuscle.flatMap { overrides in
            let relevant = involvedMuscleIds.compactMap { overrides[$0] }
            return relevant.isEmpty ? nil : relevant.min()
        }
        let relevant = [readinessNeural.map(Double.init),
                        readinessSpinal.map(Double.init),
                        readinessMuscular.map(Double.init),
                        perMuscleValues.map(Double.init)]
            .compactMap { $0 }
        let readiness = (relevant.min() ?? 100.0).clamped(to: 0.0...100.0)
        switch readiness {
        case ..<Double(readinessSevereThreshold):
            let t = (readiness / Double(readinessSevereThreshold)).clamped(to: 0.0...1.0)
            return (readinessMinFactor + (readinessSevereFactor - readinessMinFactor) * t)
                .clamped(to: readinessMinFactor...readinessSevereFactor)
        case ..<Double(readinessReductionThreshold):
            let t = ((readiness - Double(readinessSevereThreshold)) /
                (Double(readinessReductionThreshold) - Double(readinessSevereThreshold))).clamped(to: 0.0...1.0)
            return (readinessSevereFactor + (1.0 - readinessSevereFactor) * t)
                .clamped(to: readinessSevereFactor...1.0)
        case 90...:
            return readinessRecoveryFactor
        default:
            return 1.0
        }
    }

    static func buildReadinessReason(factor: Double, readinessValue: Int?) -> String {
        guard factor != 1.0 else { return "" }
        if factor < 0.85, let v = readinessValue {
            return "Readiness \(v) · −\((Int((1 - factor) * 100)))%"
        } else if factor < 1.0 {
            return "Readiness \(readinessValue ?? 0)"
        } else {
            return "Readiness \(readinessValue ?? 0) · +\((Int((factor - 1) * 100)))%"
        }
    }
}
